pyqtorch/gpsr.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import annotations
+
+from logging import getLogger
+from typing import Any, Tuple
+
+import torch
+from torch import Tensor, no_grad
+from torch.autograd import Function
+
+import pyqtorch as pyq
+from pyqtorch.analog import Observable
+from pyqtorch.circuit import QuantumCircuit
+from pyqtorch.parametric import Parametric
+from pyqtorch.utils import inner_prod, param_dict
+
+logger = getLogger(__name__)
+
+
+class PSRExpectation(Function):
+    """
+    Describe PSR
+    """
+
+    @staticmethod
+    @no_grad()
+    def forward(
+        ctx: Any,
+        circuit: QuantumCircuit,
+        observable: Observable,
+        state: Tensor,
+        param_names: list[str],
+        *param_values: Tensor,
+    ) -> Tensor:
+        ctx.circuit = circuit
+        ctx.observable = observable
+        ctx.param_names = param_names
+        ctx.state = state
+        values = param_dict(param_names, param_values)
+        ctx.out_state = circuit.run(state, values)
+        ctx.projected_state = observable.run(ctx.out_state, values)
+        ctx.save_for_backward(*param_values)
+        return inner_prod(ctx.out_state, ctx.projected_state).real
+
+    @staticmethod
+    def backward(ctx: Any, grad_out: Tensor) -> Tuple[None, ...]:
+        param_values = ctx.saved_tensors
+        values = param_dict(ctx.param_names, param_values)
+        grads_dict = {k: None for k in values.keys()}
+        shift = torch.tensor(torch.pi) / 2.0
+
+        for op in ctx.circuit.flatten():
+            if isinstance(op, Parametric) and isinstance(op.param_name, str):
+                spectrum = torch.linalg.eigvalsh(op.pauli).reshape(-1, 1)
+                spectral_gap = torch.unique(
+                    torch.abs(torch.tril(spectrum - spectrum.T))
+                )
+                spectral_gap = spectral_gap[spectral_gap.nonzero()]
+                assert (
+                    len(spectral_gap) == 1
+                ), "PSRExpectation only works on single_gap for now."
```

Comment:
The `eigvals` can be stored in the gate itself right? No need to compute this during the backward pass?

Reply:
Also, I feel like it would be best to control the applicability of `PSR` before the circuit execution, and not during the `backward` pass. E.g., each differentiation mode could have a set of supported operations, and if a circuit is called with one gate that falls outside the supported operations for the requested differentiation mode it immediately breaks. Does that make sense?

Reply:
Hey @jpmoutinho, we actually plan to quickly add the multi gap case in a very soon MR, so this check will disappear quickly. It is mostly here so @vytautas-a and myself can pick up the basic logic for `PSR` and our tasks. I will add later the methods `eigenvalues` and `eigenvalues_generator` for all the blocks similar to `qadence` in another MR.